java/src/main/native/src/row_conversion.cu

```diff
@@ -304,20 +334,600 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+          // lead thread in each warp writes data
+          auto const validity_write_offset =
+              validity_data_row_length * (relative_row + i) + relative_col / 8;
+          if (threadIdx.x % warp_size == 0) {
+            *reinterpret_cast<int32_t *>(&this_shared_block[validity_write_offset]) = validity_data;
+          }
+        }
+      }
+    }
+
+    // make sure entire block has finished copy
+    group.sync();
+
+    auto const output_data_base =
+        output_data[block.batch_number] + validity_offset + block.start_col / 8;
+
+    // now async memcpy the shared memory out to the final destination 4 bytes at a time since we do
+    // 32-row chunks
+    auto const row_bytes = util::div_rounding_up_unsafe(num_block_cols, 8);
+    auto const chunks_per_row = util::div_rounding_up_unsafe(row_bytes, 8);
+    auto const total_chunks = chunks_per_row * block.num_rows();
+    auto &subset_barrier =
+        shared_block_barriers[validity_block % NUM_VALIDITY_BLOCKS_PER_KERNEL_LOADED];
+    auto const tail_bytes = row_bytes % 8;
+
+    for (auto i = threadIdx.x; i < total_chunks; i += blockDim.x) {
+      // determine source address of my chunk
+      auto const relative_row = i / chunks_per_row;
+      auto const col_chunk = i % chunks_per_row;
+      auto const relative_chunk_offset = col_chunk * 8;
+      auto const output_dest =
+          output_data_base + row_offsets[relative_row + block.start_row] + relative_chunk_offset;
+      auto const input_src =
+          &this_shared_block[validity_data_row_length * relative_row + relative_chunk_offset];
+
+      if (tail_bytes > 0 && col_chunk == chunks_per_row - 1)
+        cuda::memcpy_async(output_dest, input_src, tail_bytes, subset_barrier);
+      else
+        cuda::memcpy_async(output_dest, input_src, cuda::aligned_size_t<8>(8), subset_barrier);
+    }
+  }
+
+  // wait for last blocks of data to arrive
+  for (int validity_block = 0;
+       validity_block < blocks_remaining % NUM_VALIDITY_BLOCKS_PER_KERNEL_LOADED;
+       ++validity_block) {
+    shared_block_barriers[validity_block].arrive_and_wait();
+  }
+}
+
+/**
+ * @brief copy data from row-based format to cudf columns
+ *
+ * @param num_rows total number of rows in the table
+ * @param num_columns total number of columns in the table
+ * @param shmem_used_per_block amount of shared memory that is used by a block
+ * @param row_offsets offset to a specific row in the input data
+ * @param output_data pointers to column data
+ * @param col_sizes array of sizes for each element in a column - one per column
+ * @param col_offsets offset into input data row for each column's start
+ * @param block_infos information about the blocks of work
+ * @param input_data pointer to input data
+ *
+ */
+__global__ void copy_from_rows(const size_type num_rows, const size_type num_columns,
+                               const size_type shmem_used_per_block, const size_type *row_offsets,
```

Comment:
As above with the `row_offsets` as an iterator.  Or is the case that in from_rows, it's the `col_offsets` that are the constants?

Reply:
Done